### ISSUE TITLE
systemd-resolved: provide a way to limit requests for non-local domains to only DNSSEC-capable servers

### ISSUE THREAD

**Issue by mikken** [systemd/systemd#8236]
### Submission type

 - Request for enhancement (RFE)

### systemd version the issue has been seen with

> 237

### Used distribution

> Arch

I'll explain my setup a little:
```
Global
         DNS Servers: 8.8.8.8
                      8.8.4.4
          DNSSEC NTA: 10.in-addr.arpa
                      16.172.in-addr.arpa
<... the rest if the list skipped ...>

Link 2 (eno1)
      Current Scopes: DNS LLMNR/IPv4
       LLMNR setting: yes
MulticastDNS setting: no
      DNSSEC setting: no
    DNSSEC supported: no
         DNS Servers: 192.168.0.6
                      192.168.0.19
          DNS Domain: example1.local
                      example2.local
```

As you see I have Google DNS as global servers with `DNSSEC=yes` and local servers with `DNSSEC=no`.
Local domains are resolved through local servers, they are used for search, too.

Everything else should not be requested from local DNS servers since they are not DNSSEC-capable. I can't achieve this with current features `systemd-resolved` has. Please provide a way to do so, since local servers answer faster and I get non-authenticated answers for signed domains.

**Comment by ott** [comment]
But can't systemd-resolved just request the DNSSEC related RRs and do the validation itself? It would be more secure than to query Google Public DNS which just sets a bit and likely also faster. I don't understand this feature request. @mikken Can you please clarify it?

**Comment by ott** [comment]
I can confirm that systemd-resolved correctly handles answers that contain RR with TTL of 0. I think it is out of scope of a stub resolver to manipulate DNS answers. It's better to use a proper recursive resolver for that. So I'm suggesting to @poettering  to close this issue.

**Comment by mikken** [comment]
About requesting related RRs, it would only work for server software that knows DNSSEC, but doesn't have validation enabled.

**Comment by ott** [comment]
You could configure systemd-resolved to validate DNSSEC itself and request the relevant RRs for the validation through the fast local server. It would be both fast and secure.

Your current idea is to rely on Google Public DNS to do the validation. As you said, your connection to Google Public DNS is slow. In addition to this, the validation result is sent without integrity protection from Google Public DNS over the Internet to systemd-resolved and can be manipulated. Effectively it is no more secure than when you did not request DNSSEC validation at all.

There are two approaches to DNSSEC validation. Either you let a resolver in your local network perform the validation, which in turn sets the AD bit of the response, and assume that network traffic on the local network cannot be manipulated or you do the validation in a stub resolver on the computer on which the software that made the DNS query runs. It seems that the second approach is preferred now, in particular if you consider RFC 7901.

So as I said, you can configure systemd-resolved to do the DNSSEC validation and let it query the necessary RRs through the local resolver. It is fast and secure.

If I'm still misunderstanding what you are trying to do, I would like to apologize and would like to ask you to explain it again to me.

**Comment by mikken** [comment]
Yes, you misunderstood. In my setup both Google and resolved perform validation. Resolved only performs validation for answers received from Google upstream server. RTT to Google isn't slow, it's about  8 ms, but local server might perform faster. As for integrity protection, I could switch to DNS over TLS and 1.1.1.1 if resolved supported it. Local DNS server will not return DNSKEY, RRSIG and those types of records. resolved cannot validate domains through it. If an answer comes faster from local server, no validation attempt will ever be performed. We have no means to tell resolved to wait for answer from DNSSEC-enabled upstream, then validate locally.

**Comment by ott** [comment]
Okay, then I understand your problem. Thank you for your patience. I'm assuming that you can't ask the operator of your local DNS server to use a different software. Is it right that you also set `DNSSEC=yes` globally?

**Comment by mikken** [comment]
Right. Small correction: if I set it globally, local resolvers will fail, and I need them to resolve local domains or for captive portals.

**Comment by ott** [comment]
So there are local domains that can only be resolved through the local resolver?

**Comment by mikken** [comment]
Yes.

**Comment by ott** [comment]
So systemd-resolved has to use Google Public DNS only for the RRs that the local resolver does not support. So there would have to be some option to tell systemd-resolved the RR types that the local resolver does not support. In addition it has to only start DNSSEC validation once you passed the captive portal because Google Public DNS is not reachable before you interacted with the captive portal?
